File: Source/Engine/src/HyperEngine/Rendering/Vulkan/Surface.cpp
/*
 * Copyright (c) 2020-2021, SkillerRaptor <skillerraptor@protonmail.com>
 *
 * SPDX-License-Identifier: MIT
 */

#include "HyperEngine/Rendering/Vulkan/Surface.hpp"

#include "HyperEngine/Core/Assertion.hpp"

#include <GLFW/glfw3.h>

namespace HyperEngine::Vulkan
{
	auto CSurface::create(const CSurface::SDescription& description) -> bool
	{
		HYPERENGINE_ASSERT_IS_NOT_NULL(description.window);
		HYPERENGINE_ASSERT_IS_NOT_EQUAL(description.instance, VK_NULL_HANDLE);

		m_instance = description.instance;

		if (glfwCreateWindowSurface(m_instance, description.window, nullptr, &m_surface) != VK_SUCCESS)
		{
			CLogger::fatal("CSurface::create(): Failed to create surface");
			return false;
		}

		return true;
	}

	auto CSurface::destroy() -> void
	{
		if (m_surface != VK_NULL_HANDLE)
		{
			vkDestroySurfaceKHR(m_instance, m_surface, nullptr);
		}
	}

	auto CSurface::surface() const noexcept -> const VkSurfaceKHR&
	{
		return m_surface;
	}
} // namespace HyperEngine::Vulkan

File: hyper_rendering/src/vulkan/core/surface.rs
/*
 * Copyright (c) 2023, SkillerRaptor
 *
 * SPDX-License-Identifier: MIT
 */

use crate::vulkan::core::instance::Instance;

use hyper_platform::window::Window;

use ash::{extensions::khr, vk, Entry};

pub(crate) struct Surface {
    handle: vk::SurfaceKHR,
    loader: khr::Surface,
}

impl Surface {
    pub(crate) fn new(window: &Window, entry: &Entry, instance: &Instance) -> Self {
        let loader = khr::Surface::new(entry, instance.handle());

        let handle = window.create_surface(entry, instance.handle());

        Self { handle, loader }
    }

    pub(crate) fn handle(&self) -> vk::SurfaceKHR {
        self.handle
    }

    pub(crate) fn loader(&self) -> &khr::Surface {
        &self.loader
    }
}

impl Drop for Surface {
    fn drop(&mut self) {
        unsafe {
            self.loader.destroy_surface(self.handle, None);
        }
    }
}
